client_test.go

```diff
@@ -448,3 +450,65 @@ func TestSubscribeWithContextAbortRetrier(t *testing.T) {
 	require.Error(t, err)
 	assert.Regexp(t, `could not connect to stream: `+http.StatusText(status), err.Error())
 }
+
+func subscribeChanWithContextUnsubscribeRace(ch chan error) {
+	srv = newServer()
+	defer srv.Close()
+
+	//Timeout for finding deadlock
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+	c := NewClient(urlPath)
+
+	eventChan := make(chan *Event)
+	err := c.SubscribeChanWithContext(ctx, "test", eventChan)
+	if err != nil {
+		ch <- err
+		return
+	}
+
+	//Simulate some messages, server automatically will close the stream
+	//after it has gone through the buffer
+	srv.Publish("test", &Event{Event: []byte("data"), Data: []byte("I've seen things you people wouldn't believe.")})
+	srv.Publish("test", &Event{Event: []byte("data"), Data: []byte("Attack ships on fire off the shoulder of Orion.")})
+	srv.Publish("test", &Event{Event: []byte("data"), Data: []byte("I watched C-beams glitter in the dark near the Tannhäuser Gate.")})
+	srv.Publish("test", &Event{Event: []byte("data"), Data: []byte("All those moments will be lost in time, like tears in rain.")})
+	srv.Publish("test", &Event{Event: []byte("finished"), Data: []byte(" ")})
+
+loop:
+	for {
+		select {
+		case ev, ok := <-eventChan:
+			if !ok {
+				srv.Close()
+				break loop
+			}
+			if bytes.Equal(ev.Event, []byte("finished")) {
+				srv.Close()
+				break loop
+			}
+		case <-ctx.Done():
+			ch <- fmt.Errorf("test deadline exceeded")
+			return
+		}
+	}
+	//Empirically set sleep to cause race condition
+	time.Sleep(time.Microsecond * 10)
+	c.Unsubscribe(eventChan)
+	ch <- nil
+}
+func TestSubscribeChanWithContextUnsubscribeRace(t *testing.T) {
+	for i := 0; i < 200; i++ {
+		ch := make(chan error, 1)
+		go subscribeChanWithContextUnsubscribeRace(ch)
+		select {
+		case err := <-ch:
+			if !assert.NoError(t, err) {
+				return
+			}
+		case <-time.After(time.Second * 2):
+			assert.Fail(t, "Should return in time, possible deadlock")
+			return
+		}
+	}
+}
```

Comment:
I tried this out without the `client.go` changes and it still passed most of the time...

There's also a data race found by `-race`:

```
WARNING: DATA RACE
Write at 0x000000c461a0 by goroutine 231:
  barney.ci/sse/v2.newServer()
      barney.ci/sse/client_test.go:57 +0x152
  barney.ci/sse/v2.subscribeChanWithContextUnsubscribeRace()
      barney.ci/sse/client_test.go:455 +0x36
  barney.ci/sse/v2.TestSubscribeChanWithContextUnsubscribeRace.gowrap1()
      barney.ci/sse/client_test.go:503 +0x33

Previous read at 0x000000c461a0 by goroutine 218:
  barney.ci/sse/v2.subscribeChanWithContextUnsubscribeRace()
      barney.ci/sse/client_test.go:487 +0xdd3
  barney.ci/sse/v2.TestSubscribeChanWithContextUnsubscribeRace.gowrap1()
      barney.ci/sse/client_test.go:503 +0x33

Goroutine 231 (running) created at:
  barney.ci/sse/v2.TestSubscribeChanWithContextUnsubscribeRace()
      barney.ci/sse/client_test.go:503 +0xe5
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      /usr/local/go/src/testing/testing.go:1742 +0x44

Goroutine 218 (running) created at:
  barney.ci/sse/v2.TestSubscribeChanWithContextUnsubscribeRace()
      barney.ci/sse/client_test.go:503 +0xe5
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      /usr/local/go/src/testing/testing.go:1742 +0x44
==================
```

Reply:
Okey, the race condition was caused by global variable usage. Normally that's okey for those tests, since they are run sequentially, but when I was spinning up goroutines that were deadlocked, then the races started to show. 

I fixed them (I hope) by making sse test server local for a given test goroutine.

Reply:
Just by running the test on my linux machine..

```
go version && go test -run TestSubscribeChanWithContextUnsubscribeRace -count 10
go version go1.22.2 linux/amd64
PASS
ok  	barney.ci/sse/v2	1.069s
```

Reply:
from a home bus, it reliably reproduces. Same go version. Good enough for me :shrug:

Reply:
Well, I lied, it doesn't reliable repro on a bus :stuck_out_tongue: 

```
-*- mode: compilation; default-directory: "~/src/barney.ci/sse/" -*-
Compilation started at Tue May  7 09:35:28

go version && go test -race -run TestSubscribeChanWithContextUnsubscribeRace -count 4
go version go1.22.2 linux/amd64
--- FAIL: TestSubscribeChanWithContextUnsubscribeRace (2.24s)
    client_test.go:516: 
        	Error Trace:	client_test.go:516
        	Error:      	Should return in time, possible deadlock
        	Test:       	TestSubscribeChanWithContextUnsubscribeRace
--- FAIL: TestSubscribeChanWithContextUnsubscribeRace (2.14s)
    client_test.go:516: 
        	Error Trace:	client_test.go:516
        	Error:      	Should return in time, possible deadlock
        	Test:       	TestSubscribeChanWithContextUnsubscribeRace
FAIL
exit status 1
FAIL	barney.ci/sse/v2	5.185s

Compilation exited abnormally with code 1 at Tue May  7 09:35:34
```

Reply:
It's probably good enough. The concern, of course, is that if there is a regression, we won't see it, because whoever hits will just retry the job :smile: